### PR TITLE
sw_engine: preventing the double application of the opacity

### PR DESCRIPTION
**Pull request by mgrudzinska**
For filled and stroked shapes with a gradient, the opacity
was applied during the composition and during the ctable update.
To prevent this, in case the composition occurred, ctable
is updated with the opacity = 255.

before:
8before

after:
8after

code:
```
    auto shape = tvg::Shape::gen();
    shape->appendRect(0, 0, 800, 800, 0, 0);
    shape->fill(255, 0, 0, 255);
    if (canvas->push(move(shape)) != tvg::Result::Success) return;

    tvg::Fill::ColorStop colorStops[2];
    colorStops[0] = {0, 0, 0, 255, 255};
    colorStops[1] = {1, 0, 0, 255, 255};


    auto shape1 = tvg::Shape::gen();
    shape1->appendRect(410, 0, 180, 180, 10, 10);

    auto fill = tvg::LinearGradient::gen();
    fill->linear(0, 0, 800, 800);
    fill->colorStops(colorStops, 2);

    shape1->fill(move(fill));
    shape1->opacity(100);
    shape1->stroke(10);

    fill = tvg::LinearGradient::gen();
    fill->linear(0, 0, 800, 800);
    fill->colorStops(colorStops, 2);

    shape1->stroke(move(fill));
    if (canvas->push(move(shape1)) != tvg::Result::Success) return;


    auto shape2 = tvg::Shape::gen();
    shape2->appendRect(610, 0, 180, 180, 10, 10);
    shape2->fill(0, 0, 255, 255);
    shape2->opacity(100);
    shape2->stroke(10);
    shape2->stroke(0, 0, 255, 255);
    if (canvas->push(move(shape2)) != tvg::Result::Success) return;


    shape1 = tvg::Shape::gen();
    shape1->appendRect(410, 210, 180, 180, 40, 40);

    shape1->fill(0, 0, 255, 255);
    shape1->opacity(100);
    shape1->stroke(30);

    fill = tvg::LinearGradient::gen();
    fill->linear(0, 0, 800, 800);
    fill->colorStops(colorStops, 2);

    shape1->stroke(move(fill));
    if (canvas->push(move(shape1)) != tvg::Result::Success) return;


    shape2 = tvg::Shape::gen();
    shape2->appendRect(610, 210, 180, 180, 40, 40);

    fill = tvg::LinearGradient::gen();
    fill->linear(0, 0, 800, 800);
    fill->colorStops(colorStops, 2);

    shape2->fill(move(fill));
    shape2->opacity(100);
    shape2->stroke(30);

    shape2->stroke(0, 0, 255, 255);
    if (canvas->push(move(shape2)) != tvg::Result::Success) return;
```

### CODE REVIEW

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 1]
```diff
@@ -131,7 +129,10 @@ struct SwShapeTask : SwTask
                 if (auto fill = sdata->strokeFill()) {
                     auto ctable = (flags & RenderUpdateFlag::GradientStroke) ? true : false;
                     if (ctable) shapeResetStrokeFill(&shape);
-                    if (!shapeGenStrokeFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:134]: In this context, stroking is always true

Contextual function calls generates more cpu instructions, you can write it one line.
shapeGenStrokeFillColors(........, (filling ? 255 : opacity), ...) goto err;
or
auto opacity2 = filling ? 255 : opacity;
shapeGenStrokeFillColors(........, opacity2, ...) goto err;

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 2]
```diff
@@ -131,7 +129,10 @@ struct SwShapeTask : SwTask
...
+                    if (stroking && filling) {
+                        if (!shapeGenStrokeFillColors(&shape, fill, transform, surface, 255, ctable)) goto err;
+                    }
+                    else if (!shapeGenStrokeFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:135]: coding convention error.

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 3]
```diff
@@ -140,6 +141,15 @@ struct SwShapeTask : SwTask
             }
         }
 
+        //Fill
+        if (validFill) {
+            auto ctable = (flags & RenderUpdateFlag::Gradient) ? true : false;
+            if (ctable) shapeResetFill(&shape);
+            if (stroking && filling) {
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:148]: auto opacity2 = (stroking && filling) ? 255 : opacity;
shapeGenStrokeFillColors(........, opacity2, ...) goto err;

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 4]
```diff
@@ -81,6 +81,7 @@ struct SwShapeTask : SwTask
             sdata->strokeColor(nullptr, nullptr, nullptr, &strokeAlpha);
         }
         bool validStroke = (strokeAlpha > 0) || sdata->strokeFill();
+        bool validFill = false;
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:84]: i think filling can be used instead.

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 5]
```diff
@@ -110,11 +111,8 @@ struct SwShapeTask : SwTask
 
         //Fill
         if (flags & (RenderUpdateFlag::Gradient | RenderUpdateFlag::Transform | RenderUpdateFlag::Color)) {
-            auto fill = sdata->fill();
-            if (fill) {
-                auto ctable = (flags & RenderUpdateFlag::Gradient) ? true : false;
-                if (ctable) shapeResetFill(&shape);
-                if (!shapeGenFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
+            if (sdata->fill()) {
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:114]: if (sdata->fill()) filling = true;
else shapeDelFill(&shape);

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 6]
```diff
@@ -110,11 +111,8 @@ struct SwShapeTask : SwTask
 
         //Fill
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:112]: Update comment: => Figuring out Filling in advance.

**Review by hermet — CHANGES_REQUESTED** [review]
Please check comments.

**Comment by mgrudzinska** [comment]
@hermet please check it after #606 and #607 are merged. here are only two lines changed if you compare to the file after its refactoring in these two prs.

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 7]
```diff
@@ -92,46 +88,43 @@ struct SwShapeTask : SwTask
...
+            if (visibleStroke) {
                 shapeResetStroke(&shape, sdata, transform);
                 if (!shapeGenStrokeRle(&shape, sdata, transform, clipRegion, bbox, mpool, tid)) goto err;
-                stroking = true;
 
                 if (auto fill = sdata->strokeFill()) {
                     auto ctable = (flags & RenderUpdateFlag::GradientStroke) ? true : false;
                     if (ctable) shapeResetStrokeFill(&shape);
-                    if (!shapeGenStrokeFillColors(&shape, fill, transform, surface, opacity, ctable)) goto err;
+                    if (!shapeGenStrokeFillColors(&shape, fill, transform, surface, visibleFill ? 255 : opacity, ctable)) goto err;
```
Comment (mgrudzinska) [inline src/lib/sw_engine/tvgSwRenderer.cpp:127]: changed: ```visibleFill ? 255 : opacity```

File: src/lib/sw_engine/tvgSwRenderer.cpp [hunk 8]
```diff
@@ -92,46 +88,43 @@ struct SwShapeTask : SwTask
...
+            if (visibleFill || visibleStroke) {
                 shapeReset(&shape);
                 if (!shapePrepare(&shape, sdata, transform, clipRegion, bbox, mpool, tid)) goto err;
-                if (renderShape) {
-                    /* We assume that if stroke width is bigger than 2,
-                       shape outline below stroke could be full covered by stroke drawing.
-                       Thus it turns off antialising in that condition.
-                       Also, it shouldn't be dash style. */
-                    auto antiAlias = (strokeAlpha == 255 && strokeWidth > 2 && sdata->strokeDash(nullptr) == 0) ? false : true;
-                    if (!shapeGenRle(&shape, sdata, antiAlias, clips.count > 0 ? true : false)) goto err;
```
Comment (hermet) [inline src/lib/sw_engine/tvgSwRenderer.cpp:105]: This looks wrong, shape rle is necessary even though only the stroking is valid but you changed it to when the filling is valid....

**Review by hermet — CHANGES_REQUESTED** [review]
Tip: Refactoring / Bug fix can be separated when you request a code review...

**Comment by hermet** [comment]
@mgrudzinska when you request a review to fix a core-side bug, please make an atomic patch by less modifying as possible.

**Comment by mgrudzinska** [comment]
@hermet I make a note that these changes are in 606 and 607 and here I add only 2 lines when compare to them. sorry for making it confusing.

**Review by hermet — APPROVED** [review]

**Pull request merged**